src/main/java/org/opensrp/web/utils/ClientFormValidator.java

```diff
@@ -106,4 +111,43 @@ private void initialisePropertiesFileJsonPathReferences() {
 
         return missingPropertyFileReferences;
     }
+
+    @NonNull
+    public HashSet<String> performWidgetValidation(@NonNull ObjectMapper objectMapper, @NonNull String formIdentifier, @NonNull String clientFormContent) throws JsonProcessingException {
+        ClientForm formValidator = clientFormService.getMostRecentFormValidator(formIdentifier);
+        HashSet<String> fieldsMap = new HashSet<>();
+        if (formValidator != null && formValidator.getJson() != null && !((String) formValidator.getJson()).isEmpty()) {
+            JsonWidgetValidatorDefinition jsonWidgetValidatorDefinition = objectMapper.readValue((String) formValidator.getJson(), JsonWidgetValidatorDefinition.class);
+            JsonWidgetValidatorDefinition.WidgetCannotRemove widgetCannotRemove = jsonWidgetValidatorDefinition.getCannotRemove();
+            if (widgetCannotRemove != null && widgetCannotRemove.getFields() != null && widgetCannotRemove.getFields().size() > 0) {
+                JsonNode jsonNode = objectMapper.readTree(clientFormContent);
+                fieldsMap.addAll(widgetCannotRemove.getFields());
+
+                Iterator<Map.Entry<String, JsonNode>> iterator = jsonNode.fields();
+                while (iterator.hasNext()) {
+                    Map.Entry<String, JsonNode> jsonFormField = iterator.next();
```

Comment:
I don't think we need to loop through all the steps & fields in case the `cannot_remove` fields are few in number. Therefore I did not see the need to use the JsonPath expressions which would return all the field keys from all the steps & then we would have to loop through the keys to confirm that required keys are still on the form